### PR TITLE
Techdocs: allow custom collation of mkdocs search document

### PR DESCRIPTION
**Pull request by alexef**
Follow up of: https://github.com/backstage/backstage/pull/20433

Original issue: https://github.com/backstage/backstage/issues/17244

After getting to use it and starting to generate the search index (including tags) [using material plugin](https://github.com/backstage/mkdocs-techdocs-core/pull/202) I realized the previous fix only partially solves our problem.

In the current code state, a Component can have tags which can be added to _all docs in the component documentation_. However, using Front Matter annotations, _an individual techdocs page could have tags_ - unique to itself.

This PR enables that, by allowing clients to override the document transformer.

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [ ] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [ ] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

**Comment by alexef** [comment]
hey @freben can you have a look at this follow up pr?

**Comment by alexef** [comment]
up!

**Comment by alexef** [comment]
hey @freben / @awanlin - any chance you can consider this?

**Review by awanlin — COMMENTED** [review]
Overall this makes sense to me, the biggest gap I have is how does one use this with the new backend system? I don't see any extension point 🤔

**Comment by awanlin** [comment]
Hi @emmaindal, would you be able to help review this please?

**Comment by alexef** [comment]
not stale.

**Comment by alexef** [comment]
hi again @awanlin 👋 
we're still interested in this change, what's the best approach to get someone to review this?

**Comment by awanlin** [comment]
I've reached out to see if I can get it reviewed by the team that owns this area 👍

File: plugins/search-backend-module-techdocs/src/collators/DefaultTechDocsCollatorFactory.ts
Comment (freben) [inline plugins/search-backend-module-techdocs/src/collators/DefaultTechDocsCollatorFactory.ts:227]: Open question with no straight answer: Would it be better to keep this triplet above the transformer call, and let the transformer return a `Partial<T>` instead? That way you only override what you want automatically without calling  `defaultTechDocsCollatorDocumentTransformer` explicitly.

Mainly this helps cover cases where the return type shape changes over time, while letting caller's code not break. For example when a new field is added.

Another pattern we've explored elsewhere is to pass in an `applyDefaults` callback to the function. Then they can choose to return `{ ...applyDefaults(=, myOverride: 1 }` or whatnot. That gives the same benefits, while also not requiring a separate export to be made for the default implementation (keeps the root api a bit cleaner)

No blocker from my end. Just thinking out loud.

**Review by freben — APPROVED** [review]
Approving just to unblock, but would be good to have the area maintainers weigh in

File: plugins/search-backend-module-techdocs/src/collators/DefaultTechDocsCollatorFactory.ts
Comment (alexlorenzi) [reply plugins/search-backend-module-techdocs/src/collators/DefaultTechDocsCollatorFactory.ts:227]: Yeah, I'm tempted to agree with this thought. Although the Entity Transformer hasn't changed, this setup is still a bit brittle to changes and requires every implementation of the Transformer to return all the properties even if they're only interested in changing 1 value. Although it's not a huge type, there's still overhead (for example maybe the implementation forgets to escape the title)...

I'd opt for the transformer returning a Partial type and move the 'defaults' back into this file so it the near to where they're being used, instead of spreading that logic across multiple files.

However, I do like that this implementation follows the same pattern so it's consistent. I think we need to make the same update to the `TechDocsCollatorEntityTransformer` and deprecate the `defaultTechDocsCollatorEntityTransformer` (probably in another PR) so that we bring these two parts into alignment.

@alexef Do you have the time to make the update to this PR?

**Review by alexlorenzi — CHANGES_REQUESTED** [review]

File: plugins/search-backend-module-techdocs/src/collators/DefaultTechDocsCollatorFactory.ts
Comment (alexef) [reply plugins/search-backend-module-techdocs/src/collators/DefaultTechDocsCollatorFactory.ts:229]: @alexlorenzi PTAL

**Comment by freben** [comment]
@alexlorenzi This one's ready for another look I think!

**Comment by alexef** [comment]
not stale, still waiting on another review :(

**Comment by freben** [comment]
@alexef in the mean time, would you mind re-running the api reports? The api-report.md file is gone and replaced with a different one.

**Comment by alexef** [comment]
on it

File: plugins/search-backend-module-techdocs/src/collators/DefaultTechDocsCollatorFactory.ts
Comment (alexlorenzi) [inline plugins/search-backend-module-techdocs/src/collators/DefaultTechDocsCollatorFactory.ts:113]:
```suggestion
    this.entityTransformer = options.entityTransformer ?? (() => ({}));
```

File: plugins/search-backend-module-techdocs/src/collators/DefaultTechDocsCollatorFactory.ts
Comment (alexlorenzi) [inline plugins/search-backend-module-techdocs/src/collators/DefaultTechDocsCollatorFactory.ts:115]:
```suggestion
    this.documentTransformer = options.documentTransformer ?? (() => ({}));
```

File: plugins/search-backend-module-techdocs/src/collators/DefaultTechDocsCollatorFactory.ts
Comment (alexlorenzi) [inline plugins/search-backend-module-techdocs/src/collators/DefaultTechDocsCollatorFactory.ts:114]:
```suggestion
```

File: plugins/search-backend-module-techdocs/src/collators/DefaultTechDocsCollatorFactory.ts
Comment (alexlorenzi) [inline plugins/search-backend-module-techdocs/src/collators/DefaultTechDocsCollatorFactory.ts:112]:
```suggestion
```

File: plugins/search-backend-module-techdocs/src/collators/DefaultTechDocsCollatorFactory.ts
Comment (alexlorenzi) [reply plugins/search-backend-module-techdocs/src/collators/DefaultTechDocsCollatorFactory.ts:113]: Need some extra parenthesis here to return an empty object (your function was has a Void return type which is why you needed the ts-ignore and the additional `?? {}` below

File: plugins/search-backend-module-techdocs/src/collators/DefaultTechDocsCollatorFactory.ts
Comment (alexlorenzi) [inline plugins/search-backend-module-techdocs/src/collators/DefaultTechDocsCollatorFactory.ts:231]:
```suggestion
                ...this.entityTransformer(entity),
```

Comment (alexlorenzi) [reply plugins/search-backend-module-techdocs/src/collators/DefaultTechDocsCollatorFactory.ts:231]: See comment above about Void return type - now the 'default' noop function returns an empty object

File: plugins/search-backend-module-techdocs/src/collators/DefaultTechDocsCollatorFactory.ts
Comment (alexlorenzi) [inline plugins/search-backend-module-techdocs/src/collators/DefaultTechDocsCollatorFactory.ts:232]:
```suggestion
                ...this.documentTransformer(doc),
```

File: plugins/search-backend-module-techdocs/src/collators/DefaultTechDocsCollatorFactory.ts
Comment (alexlorenzi) [reply plugins/search-backend-module-techdocs/src/collators/DefaultTechDocsCollatorFactory.ts:229]: Hey @alexef - This is closer, but the original idea was to keep the content of the `defaultTechDocsCollatorDocumentTransformer` in this file and completely remove that function. 

However, since we'll still need something like the `TechDocsCollatorDocumentTransformer` so that we can type the override in the parameters, I think this is an ok solution for now. I don't want to hold you up any longer. I did make some quick suggestions just to clean up a couple of parts of the code a bit, though, if. you could address those before merging.

This whole section could do with looking at again since, as you correctly point out, there are options in the Entity Transformer (and potentially) options in the Document Transformer that just aren't overridable, and that leads to a confusing API. I'll schedule some time to look over this and build a more robust solution for overriding these values, and make sure to keep you in the loop.

**Review by alexlorenzi — APPROVED** [review]

**Comment by alexef** [comment]
updated with suggestions from @alexlorenzi 🙏

**Comment by asasmith** [comment]
I'm having some issues getting this to work. I'm not sure if there are limitations when testing this locally vs in production. I've added the following to my mkdocs.yaml:
```
theme:
  name: material
  ...
plugins:
  - search
  - techdocs-core
  - tags
```
and in a test doc I've added:
```
---
tags:
  - maintainer
```
My document transformer:
```
export const techdocDocumentTransformer: TechDocsCollatorDocumentTransformer = (doc: MkSearchDoc) => {
  console.log(doc)
}
```

I can see the doc object being logged during build but none of my docs have a `tags` key. Is there something I've missed for the setup or config?

```
node: v18.20.4
yarn: 3.8.1
cli: 0.28.1
backstage: 1.32.3
```

Any help you could give would be greatly appreciated. Thank you!
```

**Comment by awanlin** [comment]
@alexef, any chance you can help @asasmith with this?

**Comment by alexef** [comment]
@asasmith are you using material search? https://github.com/backstage/mkdocs-techdocs-core/pull/202 
I believe you should also install it locally (and in the docker image) to get the search index json that contains tags.

**Comment by asasmith** [comment]
@alexef thanks for your response! Sorry, I'm just getting back to this today. I thought that the search plugin was packaged with the mkdocs material [plugin](https://squidfunk.github.io/mkdocs-material/setup/setting-up-site-search/#built-in-search-plugin). I have added the config as shown in [#202](https://github.com/backstage/mkdocs-techdocs-core/pull/202)
```
...
plugins:
    - search
    - tags
    - techdocs-core:
        use_material_search: true
...
```
I'm still not seeing a `tags` object when logging `console.log(doc)` in my transformer. FWIW I do see the `tag` html being added to the generated doc so it would appear that at least that part of it is working correctly.

Happy to share more code examples if that would be helpful to debug but I'm not sure what I'm missing at this point. Thank you!

**Pull request merged**
